### PR TITLE
Add support for multiple parameters

### LINKED ISSUE #9: Functions should be able to take multiple arguments

### PR DESCRIPTION
**Pull request by pmer**
We had a lot of code in place for this already. The only thing left was to modify the parser.

Fixes #9.

### CODE REVIEW

**Comment by pmer** [comment]
Oops, I still need to add support for typechecking these things.

EDIT: Done.

File: stratagemScripts/examples.strata [hunk 1]
```diff
@@ -1,4 +1,10 @@
...
 let x: Int = 3 + 4 in x;
 let y: Bool = true in ();
-fn(z: Bool): Bool { z; z }(let b: Bool = true in b)
+fn(z: Bool): Bool { z; z }(let b: Bool = true in b);
+
+// Identity function for integers, strings, and a conversion from the one for ints to the one for strings.
+let id_i: Int -> Int = fn(x: Int): Int { x } in
+let id_s: String -> String = fn(x: String): String { x } in
+let i_to_s: (Int -> Int) -> (String -> String) = fn(id_i: Int -> Int): String -> String { id_s } in
+(i_to_s(id_i) == id_s)
```
Comment (pmer) [inline stratagemScripts/examples.strata]: Here's what it looks like. Check out line 9. It's a function that takes a unary int function and returns a unary string function.

File: stratagemScripts/examples.strata [hunk 2]
```diff
@@ -1,10 +1,13 @@
 let x: Int = 0 in x + 1;
 let x: Int = 3 + 4 in x;
 let y: Bool = true in ();
-fn(z: Bool): Bool { z; z }(let b: Bool = true in b);
+print(fn(z: Bool): Bool { z; z }(let b: Bool = true in b));  // true
 
 // Identity function for integers, strings, and a conversion from the one for ints to the one for strings.
 let id_i: Int -> Int = fn(x: Int): Int { x } in
 let id_s: String -> String = fn(x: String): String { x } in
 let i_to_s: (Int -> Int) -> (String -> String) = fn(id_i: Int -> Int): String -> String { id_s } in
```
Comment (pmer) [inline stratagemScripts/examples.strata]: This PR adds support for type signatures that contain higher-order functions.

File: stratagemScripts/examples.strata [hunk 3]
```diff
@@ -1,10 +1,13 @@
...
-(i_to_s(id_i) == id_s)
+(print(i_to_s(id_i) == id_s));  // true
+
+let add: Int -> Int -> Int = fn(a: Int, b: Int): Int { a + b } in
+(print(add(1, 2) == 3))  // true
```
Comment (pmer) [inline stratagemScripts/examples.strata]: This PR adds support for multi-parameter functions. Unfortunately, it does not support currying, so you can't call `add(1)(2) == 3`.

Comment (pmer) [reply stratagemScripts/examples.strata]: Here is the same expression but written in a curried manner:

```
let add: Int -> (Int -> Int) =
    fn(a: Int): Int -> Int {
    fn(b: Int): Int {
        a + b
    }} in
(print(add(1)(2) == 3))  // true
```

Notice the parentheses in the type signature for `add`. They are what makes this a curried function. We can see that add takes one `Int` and returns a unary `Int` function.

If the parentheses were missing, this PR's code would parse `add`'s type signature as a function taking two `Int`s and returning an `Int`.

All this is to say that in this PR, arrows (->) do not associate right as they do in, for instance, Haskell.

**Comment by pmer** [comment]
I am wondering if this is the right approach to multiple function parameters. In discussions, @taustin, we talked about supporting multiple function parameters as being syntactic sugar for function currying. This PR does not implement that.

I'm leaning towards re-implementing it, this time using currying, to simplify proof writing later.

Feel free to comment, merge, or close!

**Comment by taustin** [comment]
Let's chat about this in our next meeting.  I tend to shy away from programming, since it is often a stumbling block for programmers new to the language.

**Comment by pmer** [comment]
We'll skip adding multiple parameters for now, but might revisit it in the future.

**Pull request closed without merging**